src/tokens/ERC1155/presets/lootbox/ERC1155Lootbox.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import {IERC1155Lootbox, IERC1155LootboxFunctions} from "./IERC1155Lootbox.sol";
+import {ERC1155Items} from "@0xsequence/contracts-library/tokens/ERC1155/presets/items/ERC1155Items.sol";
+import {MerkleProof} from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import {IERC1155ItemsFunctions} from "@0xsequence/contracts-library/tokens/ERC1155/presets/items/IERC1155Items.sol";
+
+contract ERC1155Lootbox is ERC1155Items, IERC1155Lootbox {
+    bytes32 internal constant MINT_ADMIN_ROLE = keccak256("MINT_ADMIN_ROLE");
+
+    bytes32 public merkleRoot;
+    uint256 public boxSupply;
+
+    mapping(address => uint256) private _commitments;
+    mapping(uint256 => bool) private _claimedIdxs;
+
+    constructor() ERC1155Items() {}
+
+    /// @inheritdoc ERC1155Items
+    function initialize(
+        address owner,
+        string memory tokenName,
+        string memory tokenBaseURI,
+        string memory tokenContractURI,
+        address royaltyReceiver,
+        uint96 royaltyFeeNumerator
+    ) public virtual override {
+        _grantRole(MINT_ADMIN_ROLE, owner);
+        super.initialize(owner, tokenName, tokenBaseURI, tokenContractURI, royaltyReceiver, royaltyFeeNumerator);
+    }
+
+    /// @inheritdoc IERC1155LootboxFunctions
+    function setBoxContent(bytes32 _merkleRoot, uint256 _boxSupply) external onlyRole(MINT_ADMIN_ROLE) {
+        merkleRoot = _merkleRoot;
+        boxSupply = _boxSupply;
+    }
+
+    /// @inheritdoc IERC1155LootboxFunctions
+    function commit() external {
+        if (_commitments[msg.sender] != 0) {
+            revert PendingReveal();
+        }
+        if (balanceOf(msg.sender, 1) == 0) {
+            revert NoBalance();
+        }
+        _burn(msg.sender, 1, 1);
+        uint256 revealAfterBlock = block.number + 1;
+        _commitments[msg.sender] = revealAfterBlock;
+
+        emit Commit(msg.sender, revealAfterBlock);
+    }
+
+    /// @inheritdoc IERC1155LootboxFunctions
+    function reveal(address user, BoxContent calldata boxContent, bytes32[] calldata proof) external {
+        uint256 revealIdx = getRevealId(user);
+        bytes32 leaf = keccak256(abi.encode(revealIdx, boxContent));
+
+        if (!MerkleProof.verify(proof, merkleRoot, leaf)) {
+            revert InvalidProof();
+        }
+
+        delete _commitments[user];
+        _claimedIdxs[revealIdx] = true;
+
+        for (uint256 i = 0; i < boxContent.tokenAddresses.length; i++) {
+            IERC1155ItemsFunctions(boxContent.tokenAddresses[i]).mint(
```

Comment:
Should probably use batchMint which is native to 1155

Reply:
We would need to run some extra logic before to group same tokenAddresses

Reply:
<img width="881" alt="Screenshot 2024-12-19 at 5 47 33 PM" src="https://github.com/user-attachments/assets/41095b16-f4e3-4818-89f6-87dad45954b5" />

Or smth like this?